Add reading HSS from arbitrary files

colinmitchell-ggl opened the pull request:
Add reading HSS from arbitrary files & allow HSS from multiple sources

Expose HSS filereader functions for reading/writing files

For files, allow more lax conditions on number of HSS required

Signed-off-by: Colin Mitchell <colinmitchell@google.com>

rminnich commented on pkg/hsskey/hsskey.go:
```diff
@@ -148,31 +148,40 @@ func getHssFromIpmi(verbose bool, verboseDangerous bool) ([][]uint8, error) {
-		// If it successfully reads from IPMI and the list is not empty, return the result.
-		return hssList, nil
+	if err != nil || len(hssList) == 0 {
+		log.Printf("Failed to get HSS key from IPMI: %v", err)
```
Printf from package is generally a bad idea; can you return a wrapped error?

colinmitchell-ggl replied:
Willing to take advice on the go paradigm of returning errors. The existing methodology for HSS retrieval is to throw all retrieval functions here and each are indiscriminately tried. Older use-cases got HSS _only_ from IPMI OR _only_ from eeprom. Newer use-cases will have 1 set of HSS from multiple sources.

If 1 source succeeds and one fails, I believe we want unlocking to continue. Our return structure is pretty strict with return value IFF err==nil. We also want platforms to share a common uinit, so separating tried protocols I don't think is ideal.  

I'd think a print here would be fine as a med-term solution. The boot must continue.

colinmitchell-ggl replied:
Marking resolved. Switched to a io.WriteString from an argument.

rminnich commented on pkg/hsskey/filereader.go:
```diff
@@ -156,3 +159,75 @@ func getHssFromFile(verbose bool, verboseDangerous bool, filePaths []string) ([]
+// Each HSS key is expected to be 64 bytes long and has a checksum for validation.
+// HSS that fail to validate are not written to the file.
+// Function returns error if no HSS are written or writing can no longer continue.
+func WriteHssToFile(verbose bool, verboseDangerous bool, file *os.File, hss [][]byte) error {
```
rather than two verbose bools, with the 4 combinations that apply, would a loglevel make more sense?

colinmitchell-ggl replied:
Oh, fair suggestion. I do rather like having the word dangerous be present instead of reducing it to a number due to the nature of the secrets.
But it's not a hill I'd die on.

I'll propose changing it to this (which makes more sense than original anyways. It would eliminate the unnecessary state while still keeping the verbiage. But if you'd prefer a logLevel I can go with that.
		_if verbose **|| verboseDangerous** {
		       ...
			if verboseDangerous {_

rminnich replied:
up to you, but if you have four combinations of options, you need to test all four.
And if you are determined to produce output, you really ought to produce it to an io.Writer, not just call log.Printf from a package. The io.Writer can be io.Discard if you want to null data. 

But all of us who run tests locally, or peruse test output from the CI, would *really* appreciate it if you don't just directly call log.Printf in a package. I can tell you, from experience, how much trouble that is when you're trying to find a problem.

In fact, rather than bools for verbose and verbosedangers, those can be io.Writers if you are determined to have these functions do printing, and can be set to io.Discard in the default case.

colinmitchell-ggl replied:
Marking resolved. Switched to a io.WriteString from an argument. Added combination testing for writer being nil.

rminnich commented on pkg/hsskey/filereader.go:
```diff
@@ -156,3 +159,75 @@ func getHssFromFile(verbose bool, verboseDangerous bool, filePaths []string) ([]
+
+		if verbose {
+			msg := fmt.Sprintf("Writing HSS key to file=%s", file.Name())
+			if verboseDangerous {
```
for testing, it is useful to pass an io.Writer to functions such as this, so that error output can be managed. Also, having these log.Print all over the package makes it very hard to read test output when things fail. Can you change this?

colinmitchell-ggl replied:
cmiaw, is verbose flags really passed in for unit/automated testing?
I would imagine the verbose/verboseDangerous flags already provide sufficient levers to manage test output during debugging.

I do see log.Printfs in 50+ other pkg files

rminnich replied:
yes, those crept in as people moved command code to packages. 
Just because it is done does not make it right. Let's please not add more.

rminnich replied:
and, yes, you want to pass in all value of all flags, that's how you make sure tests cover all cases.

colinmitchell-ggl replied:
Added tests here to pass in flag combinations for io.Writer + dangerious bool.

rminnich commented:
I hate to be so pushy about this, but I've been in the situation of trying to scan thousands of lines of test output trying to find an error, and dealing with packages printing when they should not. We really need to stop putting prints in packages. This is easy to fix, so please please do so.

rminnich reviewed (dismissed):
please no prints in a package. It's a real headache.

colinmitchell-ggl commented:
Switched fmt.Printf to io.Writer as suggested for all contents in the package.
Kept verboseDangerous as a bool adding to the io.Writer.
Added a few test cases for variations

hugelgupf commented on pkg/hsskey/filereader.go:
```diff
@@ -129,24 +131,24 @@ func readHssFromFile(filePath string) ([][]byte, error) {
+					toOctalEscapeSequence(hss))
 			}
-			log.Print(msg)
+			io.WriteString(warnings, msg)
```
io.WriteString does not append a new-line, but log.Print does. That's why the test is failing now

colinmitchell-ggl replied:
Duh! Thanks Chris! I had tested io.Write but for some reason I skipped testing WriteString mistaking taking the \n for granted. Fixed.

hugelgupf commented on pkg/hsskey/filereader.go:
```diff
@@ -156,3 +158,74 @@ func getHssFromFile(verbose bool, verboseDangerous bool, filePaths []string) ([]
+
+		n, err := file.Write(hssKey)
+		if n != len(hssKey) {
+			return fmt.Errorf("Failed to write entire HSS key to file %s", file.Name())
```
Don't capitalize error strings, here and everywhere else in the file: https://google.github.io/styleguide/go/decisions#error-strings

colinmitchell-ggl replied:
Thanks, fixed. Capitalizing all io.WriteString messages and lowercases all fmt.Error.

Fusxfaranto commented on pkg/hsskey/hsskey.go:
```diff
@@ -116,12 +116,12 @@ func getHssFromIpmi(verbose bool, verboseDangerous bool) ([][]uint8, error) {
 
-		if verbose {
+		if warnings != nil {
 			msg := fmt.Sprintf("HSS Entry: Id=%s", id)
 			if verboseDangerous {
 				msg = msg + fmt.Sprintf(",Seed=%x", hss)
 			}
-			log.Print(msg)
+			io.WriteString(warnings, msg+"\n")
```
Since you've changed the calling code to pass in `warnings` regardless of verbosity flags, this might be too spammy in boot logs.  I would suggest gating the whole log here on `verboseDangerous`, like you've done with similar logging in `GetHssFromFile`.

colinmitchell-ggl replied:
Fair, good catch. Updated, ptal
